framework/contrib/hit/parse.cc

```diff
@@ -997,6 +999,181 @@ explode(Node * n)
+  }
+  catch (...)
+  {
+    return false;
+  }
+}
+
+Formatter::Formatter() : canonical_section_markers(true), line_length(100), indent_string("  ") {}
+
+Formatter::Formatter(const std::string & fname, const std::string & hit_config)
+{
+  std::unique_ptr<hit::Node> root(hit::parse(fname, hit_config));
+  if (root->find("format/indent_string"))
+    indent_string = root->param<std::string>("format/indent_string");
+  if (root->find("format/line_length"))
+    line_length = root->param<int>("format/line_length");
+  if (root->find("format/canonical_section_markers"))
+    canonical_section_markers = root->param<bool>("format/canonical_section_markers");
+
+  if (root->find("format/sorting"))
+  {
+    for (auto child : root->find("format/sorting")->children())
+    {
+      if (child->path() != "pattern")
+        continue;
+
+      if (child->find("section") && child->find("order"))
+      {
+        addPattern(child->param<std::string>("section"),
+                   child->param<std::vector<std::string>>("order"));
+      }
+    }
+  }
+}
+
+std::string
+Formatter::format(const std::string & fname, const std::string & input)
+{
+  std::unique_ptr<hit::Node> root(hit::parse(fname, input));
+
+  TokenClearer tc;
+  if (canonical_section_markers)
+    root->walk(&tc, hit::NodeType::Section);
+
+  root->walk(this, hit::NodeType::All);
+  return root->render(0, indent_string, line_length);
+}
+
+void
+Formatter::addPattern(const std::string & section, const std::vector<std::string> & order)
+{
+  _patterns.push_back({section, order});
+}
+
+void
+Formatter::walk(const std::string & fullpath, const std::string & /*nodepath*/, Node * n)
+{
+  for (auto & pattern : _patterns)
+  {
+    if (!matches(fullpath, pattern.regex, true))
+      continue;
+
+    std::vector<std::string> frontorder;
+    std::vector<std::string> backorder;
+    bool onfront = true;
+    for (auto & field : pattern.order)
+    {
+      if (field == "**")
+      {
+        onfront = false;
+        continue;
+      }
+      else if (onfront)
+        frontorder.push_back(field);
+      else
+        backorder.push_back(field);
+    }
+
+    auto nodes = n->children();
+    std::vector<Node *> fronthalf;
+    std::vector<Node *> unused;
+    sortGroup(nodes, frontorder, fronthalf, unused);
+
+    std::vector<Node *> backhalf;
+    nodes = unused;
+    unused.clear();
+    sortGroup(nodes, backorder, backhalf, unused);
+
+    std::vector<Node *> children;
+    children.insert(children.end(), fronthalf.begin(), fronthalf.end());
+    children.insert(children.end(), unused.begin(), unused.end());
+    children.insert(children.end(), backhalf.rbegin(), backhalf.rend());
+
+    for (unsigned int i = 0; i < children.size(); i++)
+      children[i] = children[i]->clone();
+
+    for (auto child : n->children())
+      delete child;
+
+    for (auto child : children)
+      n->addChild(child);
+  }
+}
+
+void
+Formatter::sortGroup(const std::vector<Node *> & nodes,
```

Comment:
This method definitely needs some doco. It's pretty difficult to understand. Looks like a quadratic algorithm? Hopefully nobody puts 10000 nodes in a tree!

Reply:
Yes - it is quadratic, partly by necessity because later sorting rules can override earlier sort priorities - also because I am lazy and it is fast enough.  I ran it with sorting on the `moose_test --dump`output (multi-meg file) and it ran in ~10s of milliseconds.

Reply:
Actually - it isn't really quadratic.  It is `n*m` where n is the number of nodes and m is the number of order patterns (which should never be bigger than ~10s).